Logging ADR (#301)
stackabletech/documentation: modules/contributor/pages/adr/ADRXYZ_logging_architecture.adoc

= Logging and Log Aggregation Architecture
Felix Hennig <felix.hennig@stackable.tech>
v0.1, 2022-10-26
:status: [draft]

* Status: {status}

> * Deciders: Stackable Team

Reviewer: "Stackable Team" is not very specific for a long-lasting document. It might be clear to everyone on the team at the moment (because everyone was involved) but the team can change. In a couple of years, a newcomer might not know who was involved in the decision process.

Author: fixed, thanks ;)

Technical Story: https://github.com/stackabletech/issues/issues/202, https://github.com/stackabletech/issues/issues/261

== Context and Problem Statement

// Describe the context and problem statement, e.g., in free form using two to three sentences. You may want to articulate the problem in form of a question.

As a user of the platform, I have poor visibility into what is happening in different components of the platform, as well as how components influence each other, especially the operator and the product clusters it manages.

Logs usually give this insight. Currently, every pod logs to stdout with a certain default log format. For log configuration we support setting a custom log configuration file in some products.

Logs are **not persisted**, a crashed pod is difficult to investigate. Logs are **not aggregated**, investigating issues that involve multiple pods is difficult. Log **configuration is difficult** or impossible.

== Decision Drivers

* **identical configuration** - All product logs should be configurable in the same way, no matter their underlying logging framework (log4j, logback, Python logging, etc.). Every product should have a logging section in its CRD.

> * **easily configured sinks** - A log sink (i.e. elasticsearch) should only be configured in a single place for the whole platform (not in every product instance).

Author: We're doing a discovery configmap

=== Log Aggregation Architecture

**Agent role** - Vector offers multiple https://vector.dev/docs/setup/deployment/roles/#agent[roles] in which it can run. The DaemonSet is easier to deploy, and only one vector instance is running per host. But the sidecar has deeper integration with the product container and can read log files, something that we decided we need.

**Topology** - Vector offers multiple https://vector.dev/docs/setup/deployment/topologies/[topologies] and already lists pros and cons. An aggregator makes it easier to configure a sink: only one place to configure it (the aggregator has "multi-host context" as Vector calls it). It also reduces requests made to the sink, as it aggregates and buffers requests.

=== Log Sink: OpenSearch

https://opensearch.org/[OpenSearch] has been picked as the sink for all the aggregated logs. No other opens have been considered, but as the sink sits at the very top of the chain, it is easy to swap it out for a different software.

We want to make it easy for users to deploy OpenSearch, but also easy to switch out the sink.

(How and where is this configured, and how can we make it easy for the user?)

=== Logging configuration

> **Log levels** - It is common to have four or five log levels: Error, Warning, Info, Debug and - optionally - Trace. We want to support them across all products. The **default log level** should be Info.

Reviewer 2: Filtering of aggregated logs is much more convenient if the range of log levels is the same for all products. So I would exactly define the emitted "Stackable" log levels here:

```suggestion
**Log levels** - The following log levels are supported: ERROR, WARN, INFO, DEBUG, and TRACE. We support them across all products. The **default log level** is INFO.
```

The log levels should be mapped accordingly by the log agent. For instance, Superset emits logs with the levels CRITICAL, ERROR, WARNING, INFO, and DEBUG. The mapping should be:

| Superset log level | Stackable log level |
| --- | --- |
| CRITICAL | ERROR |
| ERROR | ERROR |
| WARNING | WARN |
| INFO | INFO |
| DEBUG | DEBUG |
| DEBUG | TRACE |

There is no TRACE log level in Superset, so if the user sets the desired log level to TRACE then it is actually set to DEBUG in Superset.

**stdout vs. file** - We want to have different log levels (and possibly other settings) for stdout and file output. This makes debugging easier, without also filling up the log aggregator with very chatty logs.

**Settings per logger** - Most products (especially Java based ones) support setting a log level per package. We also want to support this in our CRD.

```
logging:
  file:
    loggers:
      ROOT:
        level: INFO
      some.interesting.module:
        level: DEBUG
  stdout:
    loggers:
      ROOT:
        level: DEBUG
      my.specific.package:
        level: TRACE
      my.other.package:
        level: TRACE
```

**Settings per role** - All of these should be configurable per role and role group. Some sub-loggers are only available in certain roles.

```
spec:
  someRole:
    config:
      logging:
        ...
    roleGroups:
      default:
        logging:
          ...
      aDifferentGroup:
        logging:
          ...
```

> **Override everything** - The customer should be able to supply their own configuration file. Where this is placed depends on the product.

Reviewer 2: It should be mentioned that it is possible to provide a custom configuration file **per role and role-group**.

```
logging:
  custom:
    configMap: nameOfMyConfigMapWithTheConfigFile
```

> Setting the `custom` field will disable any configurations made in `file` and `stdout`.

Reviewer 2: Should we disable or disallow contradicting configurations? "Disable" means: Accept the configuration but emit a warning. "Disallow" means: Reject the configuration.

**Disable vector** - Vector should be optional, if the user wants to use their own logging system.

```
logging:

>   enableVectorAgent: false  # defaults to true

Reviewer 2: This configuration option makes only sense at the top level and not per role or role-group. So the global and non-global options differ.

Author: Hmm, are you sure? Every role-group is its own StatefulSet and so we can decide on a per role/group level if the sidecar should be added or not, can't we?

Reviewer 2: You are right.

```

=== Deploying the Stack

The operator deploys the Vector agent as a sidecar and deploys the logging configuration for the product.

> The aggregator and OpenSearch sink are deployed with Helm for now, with a plan to integrate this into stackablectl. _Maybe_ we build our own operators for Vector and OpenSearch in the future.

Reviewer 2: I already created a logging stack in stackablectl which deploys OpenSearch, OpenSearch Dashboards, and a Vector log aggregator.

Author: nice. Added that.

The connection from the agent to the aggregator is made with the help of a **discovery ConfigMap**. The aggregator is discoverable with a discovery ConfigMap, which is referenced in the ProductCluster with a property called `vectorAggregatorConfigMapName`, similar to `zookeeperConfigMapName` or `hdfsConfigMapName`. The property is located at the top level of the spec.

=== Log Sink: OpenSearch

https://opensearch.org/[OpenSearch] has been picked as the sink for all the aggregated logs. No other options have been considered, but as the sink sits at the very top of the chain, it is easy to swap it out for a different software.

The sink is configured in the Vector aggregator. We initially do not provide any "Stackable way" of configuring the sink, but it is easily configured by the user in a single location for the whole platform. Other sinks can be configured easily as well (https://vector.dev/docs/reference/configuration/sinks/[Vector docs for sink configuration]). See also <<deploying_the_stack>>.

**For now, setting up the aggregator and sink is the responsibility of the user**.

> === Logging configuration

Reviewer 2: The log configuration is currently described per pod but it must actually be done per container. So the configuration should be structured as follows:

```
spec:
  vectorAggregatorConfigMapName: ...
  role:
    roleGroups:
      myFirstRoleGroup:
        config:
          logging:
            enableVectorAgent: true
            containers:
              myFirstContainer:
                console:
                  levelThreshold: INFO
                file:
                  levelThreshold: WARN
                loggers:
                  ROOT:
                    level: INFO
                  another.logger:
                    level: ERROR
              mySecondContainer: ...
```

> **Log levels** - The following log levels are supported: ERROR, WARN, INFO, DEBUG, and TRACE. We support them across all products. The **default log level** is INFO. Products might have slightly different log levels, which are mapped accordingly by the agent. For instance, Superset emits logs with the levels CRITICAL, ERROR, WARNING, INFO, and DEBUG. The mapping should be:

Reviewer 2: Logback and log4j have the following log levels: TRACE, DEBUG, INFO, WARN, ERROR, **FATAL**, **OFF**
We should support the same. Then we can map the CRITICAL level which is used by Python to FATAL. Turning off logs for some loggers could also be useful.

Reviewer 2: It turned out that `OFF` is recognized as the boolean value `false` by YAML parsers. So `NONE` should be used instead of `OFF`.

[cols="1,1"]
|===
|Superset log level | Stackable log level 

|CRITICAL
| ERROR

| ERROR
| ERROR

| WARNING
| WARN

| INFO
| INFO

| DEBUG
| DEBUG

| DEBUG
| TRACE
|===

There is no TRACE log level in Superset, so if the user sets the desired log level to TRACE then it is actually set to DEBUG in Superset.

**console vs. file** - We want to have different log levels (and possibly other settings) for console (stdout) and file output. This makes debugging easier, without also filling up the log aggregator with very chatty logs.

**Settings per logger** - Most products (especially Java based ones) support setting a log level per package. We also want to support this in our CRD.

```
logging:
  file:
    loggers:
      ROOT:
        level: INFO
      some.interesting.module:
        level: DEBUG
  console:
    loggers:
      ROOT:
        level: DEBUG
      my.specific.package:
        level: TRACE
      my.other.package:
        level: TRACE

> ```

Reviewer 2: `file` and `console` are called appenders in logback. They are on the same level as loggers. Therefore the given structure cannot be directly translated to logback or log4j. I would suggest the following structure:

```
logging:
  containers:
    myContainer:
      console:
        levelThreshold: INFO
      file:
        levelThreshold: WARN
      loggers:
        ROOT:
          level: INFO
        another.logger:
          level: ERROR
```

The log level per logger can be defined in `loggers`. The appenders only allow to set a threshold for the log level.
